[PD-3027][PD-3163] Site traffic report.

almathew opened the pull request:
## What's this PR do?
Adds the Site Traffic report

## Screenshots (if appropriate)
![site-traffic-report](https://cloud.githubusercontent.com/assets/2728248/25283996/b44275a6-2683-11e7-82ed-758c1259e1c9.png)

## Where should the reviewer start?
There is a migration to add the report. Migrations are pretty out of hand this sprint...
Most changes were on the analytics API/utils side.
Minimal changes were done to get the analytics_event datasource working.

## How should this be manually tested?
Run the migrations.
Run `python manage.py make_data`.
Navigate to the report intentions page - site traffic should be an option.
Click through to the site traffic report and mess around. Stuff should work.

## Any background context you want to provide?
CSS for the table is still messed up for this PR and the next one I'm going to put up. 
I'm going to bother @mwarnsleyDE about what he thinks we should do for that. So a fix for that is coming separately.

tdphillips commented on analytics/urls.py:
```diff
@@ -13,8 +13,8 @@
-    url(r'^main/(?P<analytics_type>[\w-]+)$', 'analytics_main', name='analytics_main'),
-    url(r'^main$', 'analytics_main', name='analytics_main'),
+    url(r'^main/(?P<analytics_type>[\w-]+)/$', 'analytics_main', name='analytics_main'),
+    url(r'^main$/', 'analytics_main', name='analytics_main'),
```
I think these two characters might have gotten transposed.

almathew replied:
Good catch. Fixed.

tdphillips commented:
I ran the migrations and then `make_data`. This resulted in some weird results.

I cleared all of my job views and reran `make_data`, thinking I might have had some weird job views. I ended up with basically identical results.
![2017-04-21-155638_862x415_scrot](https://cloud.githubusercontent.com/assets/3379265/25294023/55c5f228-26ab-11e7-860a-e2891844376b.png)

almathew commented:
@mwarnsleyDE I'm not seeing what @tdphillips but I am seeing some wonkiness. I'm going to work on this on my own this weekend for sure, but if you have any insight into what might be going on that'd be great :+1:

almathew commented:
@tdphillips what url was that specifically? I thought @mwarnsleyDE had found all the areas where "Job Views" was hardcoded into the legend.

tdphillips commented:
@almathew: `/analytics/view/main/site-traffic/`

almathew commented:
@mwarnsleyDE some more info on what I'm seeing:
```Warning: Failed propType: Invalid prop `metricTotalThisPage` of type `string` supplied to `Table`, expected `number`. Check the render method of `TableContainer`.```

```Warning: Failed propType: Invalid prop `data[0].value` supplied to `Bar`. Check the render method of `BarChart`.```

And my personal favorite, this masterpiece
![screen shot 2017-04-21 at 4 05 04 pm](https://cloud.githubusercontent.com/assets/2728248/25294271/67ed5e68-26ac-11e7-933f-9ae9f59a0636.png)

tdphillips commented:
Probably just another "me" issue but this is what I get after choosing a country and highlighting a state.
![2017-04-21-160911_389x258_scrot](https://cloud.githubusercontent.com/assets/3379265/25294378/e47bb524-26ac-11e7-8813-f1721e47b3d8.png)

almathew commented:
There might be a slight ordering issue for the report dimensions, or this might just be me. Can someone run all the migrations and double check?
<img width="774" alt="screen shot 2017-04-23 at 11 25 16 am" src="https://cloud.githubusercontent.com/assets/2728248/25314751/bde3a770-2817-11e7-8213-d22791e04e72.png">

almathew commented:
Ok, all the JavaScript is fixed. 
@mwarnsleyDE turns out the ordering thing was becoming an issue, I did this by doing what we had discussed and having the Python side let the JS know what is broken. You might take a look at what I did because, while I tested everything and it seems to work, we all know what my JS skills are like.

I've "fixed" the CSS issue by removing blank columns from the footer. Now the only columns in the footer are ones that actually have data in them. I'm not certain this is a good long-term solution, but it's good enough for DEAM.
<img width="1234" alt="screen shot 2017-04-23 at 11 31 36 am" src="https://cloud.githubusercontent.com/assets/2728248/25314786/78476444-2818-11e7-8874-8eef18d110ea.png">

At this point I should've addressed all the concerns that @tdphillips and I have raised by making the changes to how the primary key/label was determined. 

So the only remaining issue is the primary dimensions specified for the "Job Locations" site analytics report. If a reviewer could double check that to confirm that it's just my database being wrong, that would be great :+1:
